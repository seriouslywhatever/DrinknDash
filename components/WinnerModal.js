import { useState, useCallback, useEffect } from 'react';
import { View, Text, FlatList, Modal, TouchableOpacity, ImageBackground, StyleSheet } from 'react-native';
import { useWebSocket } from '../WebsocketContext';
import { useTheme } from '../ThemeContext';
import I18n from 'react-native-i18n';

/**
 * Modal to send sips to users after the game ends.
 * Only seen by winners and only shows losers.
 */
export default WinnerModal = ({ losers, show, bet }) => {
    const [showModal, setShowModal] = useState(false);
    const [loserList, setLoserList] = useState([]);
    const [reward, setReward] = useState(bet); //Value of sips wagered times 2

    const { socket } = useWebSocket();
    const { theme } = useTheme(); //Provider of style values based on application theme.

    useEffect(() => {
        if (losers.length) {
            setLoserList(losers);
            setReward(bet * 2);
            setShowModal(show);
        }
    }, [losers, show]);

    const renderItem = useCallback(({ item }) => (
        <View style={[styles.listItemContainer, { backgroundColor: theme.backgroundColor }]}>
            <TouchableOpacity onPress={() => {
                setShowModal(false);
                socket.send(`WINNER ${JSON.stringify(item)} ${reward}`);
            }} style={{ flex: 1 }}>
                <Text style={{ fontSize: 18, color: theme.color }}>{item.username}</Text>
            </TouchableOpacity>
        </View>
    ), [reward]);

    return (
        <Modal
            animationType="slide"
            transparent={true}
            visible={showModal}
            onRequestClose={() => {
                setShowModal(false);
            }}>
            <ImageBackground source={theme.backgroundImage} style={styles.imageBackground}>
                <View style={{ flex: 1 }}>
                    <Text style={[styles.headerText, { color: theme.color }]}>{I18n.t('winnerText')}</Text>
                    <View style={{ alignItems: 'center' }}>
                        <Text style={[styles.loserSelectText, { color: theme.color }]}>{I18n.t('loserSelectMessage')}</Text>
                    </View>
                </View>
                <Text style={[styles.totalSipsText, { color: theme.color }]}>{I18n.t('totalSipsText')} {reward}</Text>
                <View style={{ flex: 3 }}>
                    <FlatList
                        data={loserList}
                        renderItem={renderItem}
                        keyExtractor={(item) => item.userId}
                        extraData={loserList}
                    />
                </View>
                <View style={{ flex: 1 }}>
                    <View style={[styles.buttonMercy, { backgroundColor: theme.backgroundColor }]}>
                        <TouchableOpacity onPress={() => setShowModal(false)}>
                            <Text style={[styles.textClose, { color: theme.color }]}>{I18n.t('showMercyText')}</Text>
                        </TouchableOpacity>
                    </View>
                </View>
            </ImageBackground>
        </Modal>
    );
}

const styles = StyleSheet.create({
    imageBackground: {
        marginHorizontal: '10%',
        borderWidth: 1,
        flex: 1,
        marginVertical: "45%"
    },
    headerText: {
        alignSelf: 'center',
        fontWeight: 'bold',
        fontSize: 40,
        marginBottom: '1%'
    },
    loserSelectText: {
        fontSize: 16,
        marginBottom: 20
    },
    totalSipsText: {
        fontSize: 18,
        fontWeight: 'bold',
        textAlign: 'center',
        marginVertical: '5%'
    },
    listItemContainer: {
        padding: 8,
        width: '100%',
        borderWidth: 1,
        flexDirection: "row",
    },
    buttonMercy: {
        position: "absolute",
        bottom: "40%",
        borderRadius: 12,
        padding: 10,
        alignSelf: 'center',
        width: '75%'
    },
    textClose: {
        fontSize: 24,
        textAlign: 'center'
    },

});