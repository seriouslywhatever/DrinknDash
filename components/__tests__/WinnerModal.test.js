import React from 'react';
import { Text, ImageBackground } from 'react-native';
import { render, fireEvent } from '@testing-library/react-native';
import renderer from 'react-test-renderer';
import WinnerModal from '../WinnerModal';

jest.mock('../../ThemeContext', () => ({
    useTheme: () => ({ theme: { color: 'black', backgroundImage: require('../../assets/wood.jpg') } })
}));

jest.mock('../../WebsocketContext', () => ({
    useWebSocket: () => ({ socket: { send: jest.fn() } })
}));

describe('WinnerModal', () => {
    it('renders correctly', () => {
        const losers = [
            { userId: '1', username: 'User1' },
            { userId: '2', username: 'User2' },
            { userId: '3', username: 'User3' }
        ];
        const tree = renderer.create(<WinnerModal losers={losers} show={true} bet={2} />).toJSON();
        expect(tree).toMatchSnapshot();
    });
    it('displays the correct number of losers', () => {
        const losers = [
          { userId: '1', username: 'User1' },
          { userId: '2', username: 'User2' },
          { userId: '3', username: 'User3' }
        ];
        const { getAllByText } = render(<WinnerModal losers={losers} show={true} bet={2} />);
        const loserItems = getAllByText(/User1|User2|User3/);
        expect(loserItems.length).toBe(losers.length);
      });
});  