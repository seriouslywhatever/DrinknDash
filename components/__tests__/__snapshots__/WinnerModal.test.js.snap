// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultModal renders correctly 1`] = `
<Modal
  animationType="slide"
  hardwareAccelerated={false}
  onRequestClose={[Function]}
  transparent={true}
  visible={false}
/>
`;

exports[`WinnerModal renders correctly 1`] = `
<Modal
  animationType="slide"
  hardwareAccelerated={false}
  onRequestClose={[Function]}
  transparent={true}
  visible={false}
/>
`;
